### PR TITLE
grab_posts_likers

### PR DESCRIPTION
**Pull request by miromannino**
Refactoring of code to retrieve likers from the last posts. This functionality was already there, but hidden and called only from `set_dont_unfollow_active_users`, these likers were merged together creating a single list of active users, which makes that retrieval only useful only for that functionality.

Now instead `get_active_users` calls the new function `get_posts_likers(browser, username, posts, skip_posts, boundary, logger)`. This function is basically the same as `get_active_users` but it retrieves the likers from the last `posts` posts as a list of tuples:

[(post1_url, likers_nicknames_list), (post2_url, likers_nicknames_list), ...]

In this way this list can be saved, or it can be utilized for further analyses (e.g. most loyal users, etc.)

`get_active_users` now has the same behaviour as before, but it uses `get_posts_likers` to avoid duplicated code.

I believe this functionality is very important to determine the quality of following people, without limiting the behaviour to a simple "unfollow if is not active in the last 7 photos".

### CODE REVIEW

**Comment by timgrossmann** [comment]
@miromannino Hey 👋 
Thank you very much for your PR. Those changes sound like they make the code more modular, which is really good. Also adding the possibility to use the added method to extract more information.

Before we can merge this you definitely have to add a section to the README describing the changes/additions by adding a section about `grab_post_likers`.

@sionking @uluQulu Could you give this a look, please?
I remember we had a discussion about the `get_active_users` before. This seems like a nice addition.
Thank you

**Comment by miromannino** [comment]
Thank you I updated the readme, tell me if that is ok.

File: README.md [hunk 1]
```diff
@@ -704,6 +705,39 @@ session.remove_follow_requests(amount=200, sleep_delay=600)
...
+
+By using this list you can rank the most active users, determine ghost followers and much more.
+The URL helps identifying which post the likers list is referred to (for example to save the list of likers to retrieve later on).
+
+`skip_posts` argument can be used to specify which post you want to start analysing. For example
+
+```python
+likers_posts_1_and_2 = session.grab_posts_likers(posts=2)
+likers_posts_3_and_4_and_5 = session.grab_posts_likers(posts=3, skip_posts=2)
+likers_post_6 = session.grab_posts_likers(posts=1, skip_posts=5)
```
Comment (uluQulu) [inline README.md:735]: It's a bit confusing? 🤔
Could you pass only the posts' index to be grabbed?
```python
grab_posts_likers(posts=[1,3,7])
```
It's straight forward logic and easier to understand.

You can get the index of the posts' with their links before grabbing and even not look for the `Next` (_not sure if you do_) button to be worked with consecutive posts..

---

Cheers 😁
---

File: instapy/util.py [hunk 2]
```diff
@@ -686,15 +685,24 @@ def get_active_users(browser, username, posts, boundary, logger):
...
+    active_users = [user for post_likers in posts_likers for user in post_likers[1]]
+
     # delete duplicated users
     active_users = list(set(active_users))
 
-    logger.info(
-        "Gathered total of {} unique active followers from the latest {} "
-        "posts in {} minutes and {} seconds".format(len(active_users),
-                                                    posts,
-                                                    diff_in_minutes,
```
Comment (uluQulu) [inline instapy/util.py:696]: Why you have removed diff entries from logger and also caused a PEP8 violation?

File: instapy/util.py [hunk 3]
```diff
@@ -515,160 +513,161 @@ def get_active_users(browser, username, posts, boundary, logger):
...
+
+                    if scroll_it is True:
+                        update_activity()
+
+                    if sc_rolled > 91 or too_many_requests > 1:  # old value 100
+                        print('\n')
+                        logger.info("Too Many Requests sent! ~will sleep some :>\n")
+                        sleep_actual(600)
+                        sc_rolled = 0
+                        if too_many_requests >= 1: too_many_requests = 0
```
Comment (uluQulu) [inline instapy/util.py:597]: wow 😭 one line `if` statement 👆🏼

File: README.md (hunk 1 above)
Comment (timgrossmann) [reply README.md:735]: @uluQulu What if we want to skip the first 2 posts and then grab the next 40?
You proposal sounds like we'd have to provide a list with each index on its own.
🤔

File: instapy/util.py (hunk 3 above)
Comment (timgrossmann) [reply instapy/util.py:597]: 😅 
I'd prefer to not have one liners in the code 🙊

File: instapy/util.py [hunk 4]
```diff
@@ -515,160 +513,161 @@ def get_active_users(browser, username, posts, boundary, logger):
...
             posts, message))
 
-    for count in range(1, posts + 1):
-        try:
-            sleep_actual(2)
+    posts_to_skip = skip_posts
+    for count in range(1, posts + skip_posts + 1):
+        sleep_actual(2)
+
+        if posts_to_skip == 0:
```
Comment (uluQulu) [inline instapy/util.py:525]: Actually, your code does not do lots of modifications as GitHub diffs here;
And it's cos of you have indented whole code block just to add another `if`/`else` statement;

Even though that is a proper way of managing code, it is adviced to respect long code blocks and use alternative approach(es).

E.g.,
Earlier it was,
```python
for count in range(1, posts + skip_posts + 1): 
    # a huge code block with ~150 lines
```

Your current code,
```python
for count in range(1, posts + skip_posts + 1): 
    sleep_actual(2)    
   
    if posts_to_skip == 0:
        # a huge code block with ~150 lines   [ALL LINES INDENTED 4 SPACES FORWARD]
    
    else: 
        logger.info("Skipping post...") 
        posts_to_skip -= 1
        # some other lines
```
You can do this to respect old code block and have Github diff show only your changes,
```python
for count in range(1, posts + skip_posts + 1): 
    sleep_actual(2)    
   
    if posts_to_skip != 0:
        logger.info("Skipping post...") 
        posts_to_skip -= 1
        # some other lines
        continue

    # a huge code block with ~150 lines   [NOTHING INDENTED]
```


I use that style whenever possible.

File: README.md (hunk 1 above)
Comment (uluQulu) [reply README.md:735]: @timgrossmann 
Note that, it is practically impossible.
After ~85 scrolls, the function stops you to get some rest, otherwise you won't be able to scroll further in that post.

And it is good to know that those native dialogs are not suitable for excessive usage.
So, if users give 40 posts to be grabbed, then it will hurt server requests which is not good...

And adding 5 times faster and much more loader (_~11 vs 50_) **graphql** approach for this purpose is more appropriate.

Also, I don't think it will take that much time to write the number of the posts you like to grab.

Anyways, rejecting my examples above and ready to go excessive? 
Then can use e.g.,
```python
grab_posts_likers(range=(1, 40), skip=[3,4])
```
Can get from 1 to 40 but skip 3rd and 4th posts providing full flexibility.

Comment (timgrossmann) [reply README.md:735]: Your example gives a lot of flexibility.
I totally agree! 👍

File: instapy/util.py (hunk 4 above)
Comment (timgrossmann) [reply instapy/util.py:525]: Nice explanation, early returns are awesome. And make things much cleaner

**Review by RaihanStark — APPROVED** [review]

**Review by RaihanStark — APPROVED** [review]

File: README.md (hunk 1 above)
Comment (miromannino) [reply README.md:735]: The function retrieving the last n posts likers. The entire code that was in get_active_users is now inside `grab_posts_likers`, they works the same, so making them different would create more confusion at this point. 

In other words, `grab_post_likers` has same interface as the existing `get_active_users`. It's just that I added the possibility to start from older posts.

The reason about adding skip_posts is that, despite we are more interested to the last posts, we might realize tomorrow that instead of the last 3 posts I am interested to the last 5 posts. I don't need to run the entire analysis again with posts=5 as parameter, I can just do posts=2 while I skip the first ones I already saved.

File: instapy/util.py (hunk 3 above)
Comment (miromannino) [reply instapy/util.py:597]: That is a replacement for the old code

```
too_many_requests = 0 if too_many_requests >= 1 else \
                        too_many_requests
```

You prefer that instead?

File: instapy/util.py (hunk 4 above)
Comment (miromannino) [reply instapy/util.py:525]: Totally agree with this... if that was the example you are describing. Here is different. 

Before we have:

```
for count in range(1, posts + skip_posts + 1): 
    # a huge code block with ~150 lines, let's call codeBlockX
    # some other lines, let's call codeBlockY
```

With the early exit with continue what the code would really need to be is:

```
for count in range(1, posts + skip_posts + 1): 
    sleep_actual(2)    
   
    if posts_to_skip != 0:
        logger.info("Skipping post...") 
        posts_to_skip -= 1
        # codeBlockY, the other lines
        continue

    # codeBlockX, the huge code block with ~150 lines
    # codeBloxkY, the other lines
```

which means replicating code, code where there are also things like  

```
next_button = browser.find_element_by_xpath(
                    "//a[contains(@class, 'HBoOv')]"
                    "[text()='Next']")
```

which you can quickly figure it out could easily be changed in one place but not in the replica in the future creating bugs.

tl;dr

the actual problem is that that 150 lines of code block shouldn't be there at the first place.
My intention here was modularizing, which is one of the first steps to make ~1000 lines of code functions become readable and reusable and less prone to bugs. The code you see was already in the old get_active_users, I just unwrapped the container that was merging all of the likers in a single list.

File: README.md (hunk 1 above)
Comment (uluQulu) [reply README.md:735]: @miromannino 

Set the range of posts you want to grab likers from:
`grab_posts_likers(range=(1, 5))`
will grab the likers from 1st, 2nd, 3rd, 4th, 5th posts.

Also, skip the posts you don't want to grab in-between that range;
`grab_posts_likers(range=(1, 5), skip=[2, 3])`
will not grab the likers from 2nd and 3rd posts.

It is not confusing, rather sophisticated and flexible, @miromannino.

---

I have read whole your code, I'm aware of that you have renamed `get_active_users()` to `grab_post_likers()` which I did not like at first, but it's okay.

Also, after that rename, you should not keep `get_active_users()` function just to merge the lists inside its tuples and add it directly to the `set_dont_unfollow_active_users()` method..

File: instapy/util.py (hunk 3 above)
Comment (uluQulu) [reply instapy/util.py:597]: It was,
```python
too_many_requests = 0 if too_many_requests >= 1 else too_many_requests
```
and @w4ngel added that backslash while PEPping.

That style is my favourite and is a direct assignment which eliminates the use of `else` in the new line.

But what you have done is what '_python supports but does not like_'.
I use that one liner coding inside a python REPL or IDLE and wouldn't use inside a real code.

File: instapy/util.py (hunk 4 above)
Comment (uluQulu) [reply instapy/util.py:525]: Well, @miromannino 
I have cloned your repo @master branch to write more specific;

miro_annotation 2019-01-27 230033

What you find difficulties or complain about is that little part (_14 lines_) which clicks the next button.
Hmm..


What you can do?
Add that next button clicking precedure with its triggering conditions to a new function:
```python
def click_next():
    # if not reached posts(parameter) value, continue
    if count + 1 != posts + skip_posts + 1 and count != 0:
        try:
            # click close button
            close_dialog_box(browser)

            # click next button
            next_button = browser.find_element_by_xpath(
                "//a[contains(@class, 'HBoOv')]"
                "[text()='Next']")
            click_element(browser, next_button)

        except Exception:
            logger.error('Unable to go to next profile post')
```

And call it inside 2 places as in my suggestion:
```python
for count in range(1, posts + skip_posts + 1): 
    sleep_actual(2)    
   
    if posts_to_skip != 0:
        logger.info("Skipping post...") 
        posts_to_skip -= 1
        # some other lines
        click_next()   # I AM A NEW LINE HERE
        continue

    # a huge code block with ~150 lines   [NOTHING INDENTED]
    click_next()   # I AM ALSO A NEW LINE HERE
```
So that you have smooth code and no 14+14 duplicated lines;
Note that, it's not a new style and InstaPy already has blocks like this due to certain reasons. 
It's just real world.

In future, `get_active_users()` [now you renamed to `get_posts_likers()`] function must get massive refactor and then it will be simplified much more (_e.g. getting total posts' count, likers' count, etc._).

So, it's best to concentrate on code quality and ease of usage.

File: README.md (hunk 1 above)
Comment (miromannino) [reply README.md:735]: 1) The function `get_active_users()` could be used by set_dont_unfollow_active_users() as well as other functions. Deleting `get_active_users()` or changing its interface causes dependent code to fail. 

2) `get_active_users()` can be used by `set_dont_unfollow_active_users()` as well as others, for example by a hypothetical `interact_with_active_users()`. 

3) by deleting `get_active_users()` the logic now in `get_active_users()` would be replicated in each and every function that wants to get active users. In the future `get_active_users()` could be smarter, and use also a `grab_post_commenters()`, or it could not even use `grab_post_likers()` anymore. 

This is how modular code is written.

File: instapy/util.py (hunk 3 above)
Comment (miromannino) [reply instapy/util.py:597]: I also changed 

```python
posts = posts if total_posts is None else total_posts if posts > \
                                                             total_posts \
        else posts
```

with this:

```python  
  if total_posts is not None and posts > total_posts:
        posts = total_posts
```

you might want to revert back that too.

File: instapy/util.py (hunk 4 above)
Comment (miromannino) [reply instapy/util.py:525]: looking forward to seeing this click_next() implemented.

**Comment by timgrossmann** [comment]
@miromannino Sorry to see you closing this PR 🙁 
Anything we could do to reopen and keep working on it? 

Your comment about making the code more modular is absolutely true and we definitely have to work on that.

File: README.md (hunk 1 above)
Comment (uluQulu) [reply README.md:735]: @miromannino 
_As well as other functions_"
 - No other function uses it. It means, no dependent code will fail.

"_..hypothetical `interact_with_active_users()`_"
 - That feature/method/function does not exist. Please, be real.

"_would be replicated in each and every function that wants to get active users_"
 - Since it is being used only for `set_dont_unfollow_active_users()` it will not be replicated.

"_In the future get_active_users() could be smarter, and use also a grab_post_commenters(), or it could not even use grab_post_likers() anymore_"
 - In the future, everything can happen but we're living the moment.
 - Also, I don't think active users would ever ignore [latest] posts' likers.

Again, after your changes, it is best suited as a variable cos all specifically it does as a function is,
```python
active_users = [user for post_likers in posts_likers for user in post_likers[1]] 
```
which is not more than a variable.

---

Also, @miromannino, many people may have different use cases of _InstaPy_, but what must be taken into account is the main repository, its future and not "_HOW PEOPLE HAVE CUSTOMIZED IT_".

---

Cheers 😁
---

File: instapy/util.py (hunk 3 above)
Comment (uluQulu) [reply instapy/util.py:597]: It is okay, @miromannino.
It is not a one-liner if statement as other I've mentioned.

But I really like my approach than yours, cos you have results 2 lines of code and also 
`posts + skip_posts` and then `total_posts - skip_posts` which is a bit confusing.

File: README.md (hunk 1 above)
Comment (miromannino) [reply README.md:735]: You are answering yourself. Useful features are difficult to implement because on how things are managed... in the moment. That's probably the mistake number one that precludes writing maintainable code.

**Comment by miromannino** [comment]
Well seeing where this discussion was going to, is clear how this project ended up having functions with 200+ LOC. I guess that is not a PEP-8 violation, but I saw people getting fired for things like that.

Nobody is "the" expert, it should be rather an objective open discussion with pro&cons. Things such as "you should", "It's just real world.", or debatable "it's easier" and "it's confusing", are straightforward converstation enders.

Said so, nothing against you, @timgrossmann. I like collaborating in projects, wherever there is compatibility.

**Comment by timgrossmann** [comment]
@miromannino I'm really sorry to hear that.
This community grew and still grows really fast an it is hard to embrace a culture of honesty that is also friendly and empathic.
If you've felt intimidated or offended in any way, I'm really sorry, that was never my or @uluQulu intention.

I absolutely agree to the point of having an objective and open discussion. Especially since there are a lot of different opinions for each and every logic.
I hope we can still find a way to reopen this PR and work on adapting and expanding your ideas, especially about breaking up the code to make it more modular, together.

You've got really important and reasonable points about the error proneness of huge code blocks and maintainability. We even have to get rid of several methods and make the remaining more adaptable to make the tooling less confusing for the users that only see this huge amount of different features, some of them even sounding kind of familiar or doing comparable actions.

We are all just learning and improving over time, unfortunately in the beginning I made the huge mistake that I pushed the really dirty and badly structured code and I just kept merging new PRs to get more features which now ended up in a huge codebase that needs to be cleaned up.

So I hope we can find a way to work together and take your ideas and improvements and apply them to InstaPy to make the tool better. 
Of course while discussing everything, not only with @uluQulu but also the other developers and contributors of InstaPy and hopefully we can all learn from those discussions.

Let me know what you think. You can also personally reach me on discord or through email:
contact.timgrossmann@gmail.com

File: README.md (hunk 1 above)
Comment (uluQulu) [reply README.md:735]: I am peculiarly au faut with the codebase of the InstaPy, @miromannino.
And adding a fully-capable feature as `interact_with_active_users()` would take less than an hour [at least to me].

Haplessly, brainteaser is not the implementation but design which co-exists within the spirit of a style that can't be ignored.
And yeah, there is also realtime evaluation which implementor must be able to handle.

**Comment by uluQulu** [comment]
@timgrossmann is right.
But since @miromannino came out with an untoward shaft of wit [against me], I'll reply.

>Well seeing where this discussion was going to, is clear how this project ended up having functions with 200+ LOC. I guess that is not a PEP-8 violation,..

More important than long lines of code written in previous history is their content and if they work as expected.
image
(_Stroustrup B. - Programming Principles and Practice Using C++, 2nd Edition - 2014_)

And refactoring code in regards of simplicity has a secondary priority and can be done any time [with ease].


>Nobody is "the" expert, it should be rather an objective open discussion with pro&cons. 
Things such as "you should", "It's just real world.", or debatable "it's easier" and "it's confusing", are straightforward converstation enders.

If you have noticed, I have expressed my subjective thought by explaining its objective details.
And if you find my review disturbing, then I am sorry. I shouldn't really review it.
But the main reason I HAVE reviewed it is I wanted to verify that you have not hurted the smart scroller I wrote a while back in the `get_active_users()` function which auto-detects and stops scrolling when all user's data have been loaded that does not work on prediction but accuracy sending no any single extra scroll request to the server.


>I like collaborating in projects, wherever there is compatibility.

You are using internal functions externally and expect _InstaPy_ project not make internal changes?
Even sometimes external functions are required to be changed.

> ...but I saw people getting fired for things like that.

Interesting.

---

Voodoo 
---

**Comment by sionking** [comment]
@uluQulu @timgrossmann @miromannino 
I must say here that I feel exactly the same as @miromannino, it seems that @uluQulu just try to move this project in his way, king of my way or no way, nothing that is not accepted by him will end up with a long and frustrating discussion, trying to make you agree with him or get frustrated during this.

I sometimes found myself spending hours on one simple variable that make not much of a different.
This is not the way it should be, it is open code, right ?

**Comment by miromannino** [comment]
Thank you @timgrossmann for the nice answer. I am glad you have a good vision of the project, it's absolutely understandable, code initially has some adjusting phases. It's an iterative process.

@sionking, you see, @uluQulu clarified why he got so peculiar about the changes: it was his code. But it's understandable, jealousy is a common feeling when people try to change your code. 
It's just a big turn-off for people wanting to help.

I actually ended up the discussion after discovering that, apparently, for @uluQulu this

```python
posts = posts if total_posts is None else total_posts if posts > \
                                                             total_posts \
        else posts
```

is to prefer to

```python  
  if total_posts is not None and posts > total_posts:
        posts = total_posts
```

Me and other people actually found that very interesting (or rather entertaining)...

**Comment by uluQulu** [comment]
Yes, @miromannino 
You are very right.
image
As I said, it is not important, 
image
But I wrote it as
```python
posts = posts if total_posts is None else total_posts if posts > total_posts else posts
```
It's just @w4ngel added backslashes for PEP8 compliance that made it 2 liner.


And I HAVE seen people deprecating existing logic UNINTENTIONALLY [or so..], e.g. @sionking is one of them.

Regarding the example I've mentioned, if you are going to offer a better solution for the smart scroller there, then go for it but please, don't try to call me jealous - "_jealousy is a common feeling when people try to change your code_" even tough you might have that common feeling.
If I were a bit of jealous, @miromannino, I would unlikely add ~15k lines into _InstaPy_.

What I care the most about OSS is preserving author rights and hate it when someone acts oppositely.


@sionking, that's a pity to read, 
>I must say here that I feel exactly the same as @miromannino, it seems that @uluQulu just try to move this project in his way, king of my way or no way, nothing that is not accepted by him will end up with a long and frustrating discussion, trying to make you agree with him or get frustrated during this.
I sometimes found myself spending hours on one simple variable that make not much of a different.
This is not the way it should be, it is open code, right ?

I'd afraid your hatred against me will be lost someday since we've collaborated together successfully for quite some time, and a few little problems in your other PR moved you into those thoughts.

When your code has had problems [in my point of view] once I viewed them, I just told them.  
You should argue my review and the code [that I reviewed] rather than me myself.

My name- **Şəhriyar**, in persian stands for "_King of the kings_", @sionking.
But it's not middle ages, you know.

---

Cheers 😁
---

**Comment by w4ngel** [comment]
Guys,

This back and forth isn’t really needed.

In @uluQulu’s defence I understand he was just trying to review the code and never had any bad intentions.
However, I also see that his language and grammar may seem like he’s quite aggressive in his nature.

We’re all here to help make the project better and to contribute to it. Moments like this are what put people off helping the project. 

@miromannino would it be possible to re-open this PR, and have someone else review? If after the review is approved the code is merged, any of the other contributors feel like it could be better from an efficiency point of view, they are more than welcome to make those changes. If you don’t want to do this, that is fine as well but can I suggest we all move on rather than commenting consistently on this PR about who knows best, at the moment this is just going to go back and forth. It’s useless, a waste of our time, your time and frankly immature.

**Comment by timgrossmann** [comment]
@uluQulu, @sionking @miromannino 
As @w4ngel mentioned, `Moments like this are what put people off helping the project`and I'm glad that this PR has already been closed. Discussions like this make everyone lose. If devs interested in contributing see the discussions and behaviour here, they definitely won't be interested in contributing to InstaPy.

May I please ask you to keep the comments very short and on point from now on?
After all, this is technical work, heavy emotions and personal attacks/defences do not belong here.
The time invested in this elongated discussions is better suited somewhere else.

@miromannino If you're interested in what @w4ngel proposed with re-opening this PR and having someone else to help review and work on it, could you please open another PR with the same additions?
As mentioned, it would be nice to make this addition available for all users of InstaPy. 

I mean it's only the default of GitHub, but I still think those are very good guidelines and I would be more than happy if we could make this https://github.com/timgrossmann/InstaPy/blob/master/CODE_OF_CONDUCT.md the standard for people working together as a team on InstaPy.

Could we please agree on that? :v:
